### DISCUSSION TITLE
Generating Debug Certificate

### DISCUSSION THREAD

**Discussion by mrbenp** [codefori/vscode-ibmi#2169] [Q&A]
👉🏻 Can the debug certificate be generated outside of VSCODE.  I don't have ALLOBJ authority, and my admins will not give me a temp id will obj authority.  Is there a command I can give the admins to run on the green screen that will generate the certificate?

<hr />

⚠️ **REMOVE THIS LINE AND ANY SENSITIVE INFORMATION BELOW!** ⚠️

|Context|Version|
|-|-|
|Code for IBM i version|2.11.0|
|Visual Studio Code version|1.91.1|
|Operating System|win32_x64|

<details>
<summary>Active extensions</summary>

```
AngularDoc for Visual Studio Code (angulardoc-vscode): 6.1.3
AngularJS Template Autocomplete (angularjs-template-autocomplete): 1.0.13
COBOL (cobol): 7.7.9
Code Coverage for IBM i (code-coverage-ibmi): 0.1.7
Code for IBM i User Spooled Files (vscode-ibmi-UserSpooledFiles): 0.0.5
Code for IBM i Walkthroughs (vscode-ibmi-walkthroughs): 0.5.0
Configuration Editing (configuration-editing): 1.0.0
Db2 for IBM i (vscode-db2i): 1.0.0
ESLint (vscode-eslint): 3.0.10
Emmet (emmet): 1.0.0
Error Lens (errorlens): 3.20.0
Extension Test Runner (extension-test-runner): 0.0.11
Git (git): 1.0.0
Git Base (git-base): 1.0.0
GitHub (github): 0.0.1
IBM i Debug (ibmidebug): 2.0.1
IBM i Notebooks (vscode-ibmi-notebooks): 0.0.6
JSON Language Features (json-language-features): 1.0.0
Jira and Bitbucket (Atlassian Labs) (atlascode): 3.0.10
Live Server (LiveServer): 5.7.9
Merge Conflict (merge-conflict): 1.0.0
NPM support for VS Code (npm): 1.0.1
Node Debug Auto-attach (debug-auto-launch): 1.0.0
Red Hat Dependency Analytics (fabric8-analytics): 0.9.4
TODO Highlight (vscode-todo-highlight): 1.0.5
TypeScript and JavaScript Language Features (typescript-language-features): 1.0.0
YAML (vscode-yaml): 1.15.0
Zowe Explorer (vscode-extension-for-zowe): 2.16.3
ng1.5 components utility (vscode-angular-components-intellisense): 1.0.0
```
</details>

<hr />

<details>
<summary>Remote system</summary>

|Setting|Value|
|-|-|
|IBM i OS|V7R3M0|
|Tech Refresh|n/a|
|CCSID Origin|37|
|Runtime CCSID|37|
|Default CCSID|37|
|SQL|Enabled
|Source dates|Disabled

### Enabled features

|/QOpenSys/pkgs/bin|/usr/bin|/QSYS.lib/ILEDITOR.lib||/QSYS.LIB|/QIBM/ProdData/IBMiDebugService/bin|
|-|-|-|-|-|-|
|bash|attr|GETNEWLIBL.PGM|ILEDITOR.GETMBRINFO|QZDFMDB2.PGM|startDebugService.sh|
|chsh|iconv|||||
|grep|setccsid|||||
|ls|tar|||||
|md5sum||||||
|sort||||||
|stat||||||
</details>

**Comment by worksofliam** [comment]
@mrbenp In theory it can. You would have to [dissect our code base](https://github.com/codefori/vscode-ibmi/blob/master/src/api/debug/certificates.ts#L80) for it, since we actually do more than just generating certificates.

**Comment by sebjulliand** [comment]
@mrbenp can one of your admin connect using VSCode and perform the configuration from there? Once is enough - as soon a the certificate and the configuration file will be updated by Code for i, they will be able to start it and schedule it to start after each IPL (the SBMJOB used to start it can be found in the documentation: https://codefori.github.io/docs/developing/debug/#starting-the-debug-service-outside-of-code-for-ibm-i)

**Comment by mrbenp** [comment]
Thanks for all your input guys!  In my shop, the admins are still working from the green screen, and have no knowledge about VScode for the IBMi,   so it is not easy for me to get them to generate the certificate from VScode. Then another factor that plays in is we are all remote, and spread out across the country, so there is no admin that can walk to my computer, sign on to vscode and generate the certificate. 
It's also highly unlikely that the admins will install VScode on their machine just to generate the certificate;  with the exception of myself, the developers as well work remotely, and strictly on the green screen.  

I guess my only option for generating the certificate, is to keep asking the admins, until they relent.

One reason for posting this is, I'm curious if there are others in the same situation, and how they overcame this barrier.

**Reply by MarcoDeSenas** [reply]
Sorry to disturb the discussion but...
Wow, does it mean that your admins do not have any idea about the way you are developing applications?
If yes, IMHO they are not admins, they are operators!
They should be active members on discussions and decisions in place to organize systems nowadays. For instance, switching from physical files members sources to IFS stream files sources, setting up modern DevOps techniques, leveraging OpenSource technology, controlling applications life cycle and security, and so on.
They could learn a lot themselves if they used VScode for some of their day-to-day works, for instance developing all the tools that normal admins want to have and do not find on the web.
I cannot imagine how admins can perform their job in 2024 with only green screen. Do they ignore PASE and QShell? Do they ignore iACS? Do they ignore Navigator for i? Do they ignore SQL?

So be brave and good luck with them!

**Reply by chrjorgensen** [reply]
@MarcoDeSenas They have `bash` etc. on the system, so someone has installed some of the open source tools. So all hope is not lost.
Unfortunately the technology refresh level is not available, so we can't see how updated the system is. For debug service to work, it requires an updated Java11 - the latest PTF group SF99725 is level 33.

@mrbenp Btw, IBM i 7.3 went out of regular support in September 2023 - I hope there are plans for an upgrade to IBM i 7.5 some time in the near future... 🙏

**Reply by MarcoDeSenas** [reply]
Ha yes, and working hard to convince management not to run unsupported software (and hardware) is also one of the admins tasks...

**Comment by mrbenp** [comment]
Thanks for all the great comments.   The java version is 8.  Looks like I'm toast with getting debug to work!

**Reply by sebjulliand** [reply]
Yep, if you have Debug Service v2 and no Java 11, there's no point in trying to do anything right now.
But we'll work on a way to generate a shell script from Code for i that will allow you to reproduce the debug service's configuration process. That will give you time to ask your admins to install Java 11 😄
